php/libraries/Candidate.class.inc

```diff
@@ -699,52 +699,92 @@ class Candidate
     }
 
     /**
+    * Generates a new id in the format defined by the settings
+    *
+    * @param string   $name                  The name to display on exception
+    * @param array    $settings              The xml settings
+    * @param array    $static_dict           Example: array("siteAbbrev"=>"DCC")
+    * @param function $is_colliding_callback `bool ($needle)`
+    *
+    * @return string The id
+    */
+    private static function _generateId(
+        $name,
+        $settings,
+        $static_dict,
+        $is_colliding_callback
+    ) {
+        $generator = Candidate::idGeneratorFromXMLConfig(
+            $settings,
+            $static_dict
+        );
+
+        $is_sequential = (
+            isset($settings["generation"])
+            && $settings["generation"] == "sequential"
+        );
+
+        $result = null;
+        if ($is_sequential) {
+            if ($generator->nextNonColliding($is_colliding_callback)) {
+                $result = $generator->cur();
+            }
+        } else {
+            do {
+                $generator->nextRand();
+                $result = $generator->cur();
+            } while (!is_null($result) && $is_colliding_callback($result));
+        }
+
+        if (is_null($result)) {
+            $attempts = $generator->getAttempts();
+            $seqCount = $generator->sequenceCount();
+            $error    = "Could not generate {$name} after {$attempts} attempts. "
+                . "All {$seqCount} possible {$name} have been generated. "
+                . "Please set {$name} generator definition "
+                . "to allow more {$name} to be generated. ";
+            if (!$is_sequential) {
+                $error .= "(expected number of candidates)^3 is recommended "
+                    . "for random generation.";
+            }
+            throw new Exception($error);
+        }
+        return $result;
+    }
+    /**
     * Generates a new PSCID in the format defined by the config file
     *
-    * @param string $prefix The prefix to use for the PSCID
+    * @param array $static_dict Example: array("siteAbbrev"=>"DCC")
     *
-    * @return string Project candidate id in format defined by config.xml
+    * @return string id in format defined by config.xml
     */
-    static function _generatePSCID($prefix)
+    static function _generatePSCID($static_dict)
```

Comment:
Because the purpose is a lot different now. And I would say that `$prefix` is a terrible name for it. It might *usually* be used as a prefix but that's not what it really is. It could very well be used as a suffix as its value *was* the site abbreviation.

What it does now is take an associative array of what is "defined" as a `static` element elsewhere but is an alias for a "static" (unchanging) value.